docs/awsprivatelink.md

```diff
@@ -0,0 +1,121 @@
+allow only Hive to access the API but these are usually not acceptable by
+security focused customers.
+
+AWS provides a feature called AWS Private Link ([see doc][aws-private-link-doc]) that allows
+accessing private services in customer VPCs from another account using AWS's internal
+networking and not the Internet. AWS Private Link involves createing a VPC
+Endpoint Service in customer's account that is backed by one or more internal
+NLBS and a VPC Endpoint in service provider's account. This allows clients in
+service provider's VPC to access the NLB backed service in customer's account
+using the endpoint-endpoint service Private Link. So the internal service is
+now accessible to the service provider without exposing the service to the
+Internet.
+
+Using this same architecture, we create a VPC Endpoint Service, VPC Endpoint
+pair to create an Private Link to cluster's internal NLB for k8s API server,
+allowing Hive to access the API without forcing the cluster to publish it on
+the Internet.
+
+## Configuring Hive to enable AWS Private Link
+
+To configure Hive to support Private Link in a specific region,
+
+1. Create a VPCs in that region that can be used to create VPC Endpoints.
+   NOTE: There is a hard limit of 255 VPC Endpoints in a region, therefore you
+   will need multiple VPCs to support more cluster in that region.
+
+1. For each available availability zone in the region, create a subnet for all
+   VPCS.
+
+1. Make sure all the Hive environments (Hive VPCs) have network reachability to
+   these VPCs using peering, transit gateways, etc.
+
+1. Gather a list of VPCs that will need to resolve the DNS setup for Private
+   Link. This should atleast include the VPC of the Hive being configured, and
+   can include list of all VPCs where various Hive controllers exists.
+
+1. Update the HiveConfig to enable Private Link for clusters in that region.
+
+```yaml
+##hiveconfig
+spec:
+  awsPrivateLink:
+    ## this this is list if inventory of VPCs that can be used to create VPC
+    ## endpoints by the controller
+    vpcEndpointInventory:
+    - region: region1
+      vpcID: vpc1-region1
+      subnets:
+      - availabilityZone: region1-az1
+        subnet: vpc1-subnet1
+      - availabilityZone: region1-az2
+        subnet: vpc1-subnet3
+      - availabilityZone: region1-az2
+        subnet: vpc1-subnet3
+    - region: region1
+      vpcID: vpc2-region2
+      subnets:
+      - availabilityZone: region1-az1
+        subnet: vpc2-subnet1
+      - availabilityZone: region1-az2
+        subnet: vpc2-subnet3
+      - availabilityZone: region1-az2
+        subnet: vpc2-subnet3
+    
+    ## credentialsSecretRef points to a secret with permissions to create
+    ## resources in account where the inventory of VPCs exist.
+    credentialsSecretRef:
+      name: < hub-account-credentials-secret-name >
+
+    ## this is a list of VPC where various Hive clusters exists.
+    associatedVPCs:
```

Comment:
these are not just hive VPCs, but any VPC that will need to access the cluster api like they would have if the clsuter was on the internet. this can and should include any jump network that SREs might use to access the cluster API.

Reply:
Sounds good.